state/factory/factory.go

```diff
@@ -220,7 +227,22 @@ func (sf *factory) Start(ctx context.Context) error {
 	default:
 		return err
 	}
-	return sf.lifecycle.OnStart(ctx)
+	sf.mutex.Unlock()
+	needUnlock = false
+
+	// start the protocols
+	ctx = protocol.WithStateReaderCtx(ctx, protocol.StateReaderCtx{SR: sf})
```

Comment:
as yutong mentioned previously, we shouldn't pass so many parameters via ctx.

Reply:
ok, i find a way such that we don't need to pass SR to protocol at Start()